### PR TITLE
remove breaking changes in SyncSession.ErrorHandler

### LINKED ISSUE #4380: ClientReset error reporting should not be a breaking change
After some internal discussion, we decided that breaking semantic versioning in #4080 wasn't acceptable after all. To be as compatible with Cocoa as possible as well as to minimise changes now and later (where Client Reset might be handled entirely internally) we agreed on the following changes:

- [x] Remove `ErrorHandler.onClientResetRequired(SyncSession,ClientResetHandler handler)` method again
- [x] Move the doc from that method to somewhere sensible. Probably to `ClientResetHandler`, but we also need to document how to handle it in the `onError` javadoc.
- [x] Report ClientReset through `ErrorHandler.onError(ObjectServerErrror error)`
- [x] Given this change we should consider renaming `ClientResetHandler` back to `ClientResetRequiredError`?

### PR DESCRIPTION
**Pull request by zaki50**
fixes #4380

### CODE REVIEW

File: realm/realm-library/src/objectServer/java/io/realm/ErrorCode.java [hunk 1]
```diff
@@ -127,8 +127,8 @@ public static ErrorCode fromInt(int errorCode) {
         throw new IllegalArgumentException("Unknown error code: " + errorCode);
     }
 
-public enum Category {
+    public enum Category {
         FATAL,          // Abort session as soon as possible
-        RECOVERABLE    // Still possible to recover the session by either rebinding or providing the required information.
+        RECOVERABLE,    // Still possible to recover the session by either rebinding or providing the required information.
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/ErrorCode.java]: I'm surprised it works having a `,` on the last entry in an enum?

File: realm/realm-library/src/objectServer/java/io/realm/SyncSession.java [hunk 2]
```diff
@@ -251,13 +251,9 @@ void close() {
...
-         * @param session {@link SyncSession} this error happened on.
-         * @param error type of error.
-         */
-        void onError(SyncSession session, ObjectServerError error);
-
-        /**
-         * An error that indicates the Realm needs to be reset.
+         * <p>
+         * When the {@code error.getErrorCode()} returns {@link ErrorCode#CLIENT_RESET}, it indicates the Realm
+         * needs to be reset and {@code error} instance can be cast to {@link ClientResetObjectServerError}.
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncSession.java]: and the {@code error} can be cast to

File: realm/realm-library/src/objectServer/java/io/realm/ClientResetObjectServerError.java [hunk 3]
```diff
@@ -21,16 +21,16 @@
 /**
  * Class encapsulating information needed for handling a Client Reset event.
  *
- * @see io.realm.SyncSession.ErrorHandler#onClientResetRequired(SyncSession, ClientResetHandler) for more information
+ * @see io.realm.SyncSession.ErrorHandler#onError(SyncSession, ObjectServerError) for more information
  *      about when and why Client Reset occurs and how to deal with it.
  */
-public class ClientResetHandler extends ObjectServerError {
+public class ClientResetObjectServerError extends ObjectServerError {
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/ClientResetObjectServerError.java]: I'm not sure how I feel about this name. It seems overly long. Also consider the class has a method named `executeClientReset()`.

Some other suggestions:
- `ClientResetRequiredError`
- `ClientResetError`

Thoughts?

File: realm/realm-library/src/objectServer/java/io/realm/ErrorCode.java (hunk 1 above)
Comment (zaki50) [inline realm/realm-library/src/objectServer/java/io/realm/ErrorCode.java]: https://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.9

Java language spec says that enum accepts optional `,` before EnumBodyDeclarations.

File: realm/realm-library/src/objectServer/java/io/realm/ClientResetObjectServerError.java (hunk 3 above)
Comment (zaki50) [inline realm/realm-library/src/objectServer/java/io/realm/ClientResetObjectServerError.java]: `ClientResetRequiredError` is fine for me.

File: realm/realm-library/src/objectServer/java/io/realm/SyncSession.java (hunk 2 above)
Comment (zaki50) [inline realm/realm-library/src/objectServer/java/io/realm/SyncSession.java]: fixed

File: realm/realm-library/src/objectServer/java/io/realm/ClientResetObjectServerError.java (hunk 3 above)
Comment (zaki50) [inline realm/realm-library/src/objectServer/java/io/realm/ClientResetObjectServerError.java]: done

**Pull request merged**